File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_1.java
package org.twyszomirski.aoc.solutions;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.io.IO.println;

public class Day_1 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_1.txt"));
        println("======== Day 1 ===========");
        part_1(lines);
        part_2(lines);
    }

    public static void part_1(List<String> lines) {
        var rows = lines.stream()
                .map(it -> Arrays.stream(it.split(" ")).filter(v -> !v.isEmpty()).toList())
                .map(row -> new Pair<>(Long.valueOf(row.getFirst()), Long.valueOf(row.getLast())))
                .toList();


        var leftList = rows.stream().map(Pair::left).sorted().toList();
        var rightList = rows.stream().map(Pair::right).sorted().toList();

        var sum = 0L;
        for (int i = 0; i < leftList.size(); i++) {
            sum += Math.abs(leftList.get(i) - rightList.get(i));
        }

        System.out.println(sum);
    }

    public static void part_2(List<String> lines) {
        var rows = lines.stream()
                .map(it -> Arrays.stream(it.split(" ")).filter(v -> !v.isEmpty()).toList())
                .map(row -> new Pair<>(Long.valueOf(row.getFirst()), Long.valueOf(row.getLast())))
                .toList();


        var leftList = rows.stream().map(Pair::left).sorted().toList();
        var frequency = rows.stream().map(Pair::right).collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));

        var sum = leftList.stream().mapToLong(v -> v * frequency.getOrDefault(v, 0L)).sum();

        println(sum);
    }

    record Pair<T>(T left, T right) { }
}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_1.kt
package org.twyszomirski.aoc.solutions

import java.io.File
import kotlin.collections.mutableListOf
import kotlin.math.abs

class Day_1 {

    fun solve(){
        val lines = File("src/main/resources/input_day_1.txt").readLines()
        println("======== Day 1 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val folded = input
            .map { it.split(" ").filter { it.isNotBlank() } }
            .map { Pair(it[0].toLong(), it[1].toLong()) }
            .fold(Pair(mutableListOf<Long>(), mutableListOf<Long>()), operation = { acc, pair ->
                acc.first.add(pair.first)
                acc.second.add(pair.second)
                acc
            })
        val leftSorted = folded.first.sorted()
        val rightSorted = folded.second.sorted()

        val sum = leftSorted.zip(rightSorted).sumOf { abs(it.first - it.second) }
        println(sum)
    }

    fun part2(input: List<String>) {
        val folded = input
            .map { it.split(" ").filter { it.isNotBlank() } }
            .map { Pair(it[0].toLong(), it[1].toLong()) }
            .fold(Pair(mutableListOf<Long>(), mutableListOf<Long>()), operation = { acc, pair ->
                acc.first.add(pair.first)
                acc.second.add(pair.second)
                acc
            })

        val leftList = folded.first
        val spectrum = folded.second.groupingBy { it }.eachCount()

        val sum = leftList.sumOf { it * (spectrum[it] ?: 0) }

        println(sum)
    }
}